Diff transfer improvements

ffuugoo opened the pull request:
Minor fixes/improvements/cleanups for diff transfer related code.

Tracked in: #3477

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -544,7 +544,7 @@ impl LocalShard {
         // index *occasionally*), but the storage can handle it.
 
-        for (op_num, update) in wal.read_all() {
+        for (op_num, update) in wal.read(wal.first_index()) {
```
Why do we want to be more explicit here, isn't `read_all()` enough, fulfilling this purpose?

timvisee replied:
Makes sense!

Yes, I'd vote for a `read_all(with_acknowledged)` which defines `first_index` or the other internally, returning back the doubly ended iterator.

ffuugoo replied:
Tbh, I don't like these `with_acknowledged` methods. In my opinion, it makes *very* little difference with just doing `wal.read(wal.some_specific_index())`. Especially when we only ever use `read_all` *once* and always with `with_acknowledged = false`.

But I only *slightly* disagree with this, so I'm not too against `read_all(with_acknowledged)`.

timvisee replied:
> Especially when we only ever use read_all once and always with with_acknowledged = false.

We don't.

When resolving the diff it would use `read_all(true)`.

ffuugoo replied:
Yeah, I see now. Before we only used `read_all` in a single place. It makes sense now. 👍

timvisee commented on lib/collection/src/shards/shard.rs:
```diff
@@ -151,7 +151,10 @@ impl Shard {
         match self {
             Self::Local(local_shard) => Ok(local_shard.recovery_point().await),
+
+            // TODO: Is it good idea to return `RecoveryPoint` for `ForwardProxyShard`? 🤔
```
Maybe not, but my thought was that it's still a regular shard internally.

I guess we can assume how that we only request a recovery point if we want to recover this node, in which case it should not have a forward proxy.

So I'm fine by limiting it to be cautious, and removing this here.

timvisee commented on lib/collection/src/shards/shard.rs:
```diff
@@ -167,13 +170,16 @@ impl Shard {
         let resolve_result = match self {
             Self::Local(local_shard) => local_shard.wal.resolve_wal_delta(recovery_point).await,
+
+            // TODO: Same as `shard_recovery_point`, is it a good idea to resolve WAL delta for `ForwardProxyShard`?
```
Same as the above basically, I've only routed it here because we can.

timvisee commented on lib/collection/src/wal.rs:
```diff
@@ -110,8 +110,16 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
-    pub fn read_all(&'s self) -> impl Iterator<Item = (u64, R)> + 's {
-        self.read(self.first_index())
+    pub fn read(&self, start_index: u64) -> impl DoubleEndedIterator<Item = (u64, R)> + '_ {
+        (start_index..self.last_index()).map(move |idx| {
```
I thought the `last_index()` is inclusive. Which is interesting, because if that is the case, it means the previous implementation is broken too.

```suggestion
        (start_index..=self.last_index()).map(move |idx| {
```

Note: I'm investigating!

timvisee replied:
This is the case indeed! :exploding_head: 

```rust
    /// Last index that is still available in logical WAL.
    pub fn last_index(&self) -> u64 {
        self.wal.last_index()
    }
```

which points to:

```rust
    /// The index of the last entry
    pub fn last_index(&self) -> u64 {
        let num_entries = self.num_entries();
        self.first_index() + num_entries.saturating_sub(1)
    }
```

The `num_entries.saturating_sub(1)` means the higher bound is inclusive.

So we want to loop over that records as well:

```diff
-(start_index..self.last_index())
+(start_index..=self.last_index())
```

ffuugoo replied:
Yeah, that's probably just a bug I've introduced. Thanks. (Maybe that's why tests failed. This PR is more of idea dump atm.)

ffuugoo replied:
Fixed this, to see if it will fix CI.

timvisee commented on lib/collection/src/wal.rs:
```diff
@@ -110,8 +110,16 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
 
-    pub fn read_all(&'s self) -> impl Iterator<Item = (u64, R)> + 's {
-        self.read(self.first_index())
+    pub fn read(&self, start_index: u64) -> impl DoubleEndedIterator<Item = (u64, R)> + '_ {
```
I like that we use `DoubleEndedIterator` here now so we have a single function. But I must admit, this way I find `start_index` very confusing.

I'd rather have a `read_all(with_acknowledged: bool)` which is also doubly ended to prevent any confusion here.

I'd strongly vote for the latter because it prevents exposing unnecessary implementation details - such as record indices - to the public.

ffuugoo replied:
I think I have to agree now. I'll return the `read_all` method back. I also don't like these indices, cause they are confusing as hell.

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -14,8 +14,12 @@ pub type LockedWal = Arc<ParkingMutex<SerdeWal<OperationWithClockTag>>>;
+
     /// Map of all highest seen clocks for each peer and clock ID.
+    //
+    // TODO: Rename to `newest_clocks` (or `newest_*` something)?
```
Sure! I see that you've also used _new_ in other places, so let's follow that here then.

Please also update the file name to `clock_map_newest.json`.

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -24,11 +28,14 @@ pub struct RecoverableWal {
     /// - this WAL cannot resolve any delta below any of these clocks
+    //
+    // TODO: Rename to `oldest_available_records` or `oldest_resolvable_records`?
+    // TODO: What does "same as `highest_clocks`" note above mean?
```
It means that entries in `cutoff_clocks` are never newer than `highest_clocks`.

So, if we update cutoff, we also update `highest_clocks`.

ffuugoo replied:
> It means that entries in cutoff_clocks are never newer than highest_clocks.
> So, if we update cutoff, we also update highest_clocks.

I do get this... but I still don't see how "same as `highest_clocks`" can mean this. I'll just reword the quote above and replace this part later. :D

timvisee replied:
It notes 'at least', but I agree it is confusing.

> I'll just reword the quote above and replace this part later. :D

Thanks!

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -24,11 +28,14 @@ pub struct RecoverableWal {
     /// - this WAL has at least all these clock versions (same as `highest_clocks`)
     /// - this WAL cannot resolve any delta below any of these clocks
+    //
+    // TODO: Rename to `oldest_available_records` or `oldest_resolvable_records`?
```
Don't really like that name, but I don't have a better suggestion. I'd prefer something short.

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -81,18 +92,19 @@ impl RecoverableWal {
+            // A recovery point has every clock one higher, here we decrease it by one again
+            //
+            // TODO: Refactor `RecoveryPoint` to transfer *current* clocks instead of *next* clocks!
+            clock_tag.clock_tick -= 1;
+
+            // Advance highest clocks we've seen
+            highest_clocks.advance_clock(clock_tag);
+
+            // Advance the cutoff point
+            cutoff_clocks.advance_clock(clock_tag);
```
Nitpick, but after reading this a bunch of times I don't think we have to be overly explicit.

```suggestion
            clock_tag.clock_tick -= 1;
            highest_clocks.advance_clock(clock_tag);
            cutoff_clocks.advance_clock(clock_tag);
```

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -192,13 +204,16 @@ fn resolve_wal_delta(
+    let wal = local_wal.lock();
+
+    let delta_from = wal
+        .read(wal.first_closed_index())
+        .rev()
```
This is where I'd prefer the following:

```suggestion
        .read_all(true)
        .rev()
```

ffuugoo replied:
Hmmm. Maybe you're right. Ok, I'll return the `read_all`. 👍

timvisee reviewed (commented):
Some of my comments on the current state.

ffuugoo commented:
Resolved all previous comments, organized history, rebased on `dev`, added a few more improvements. Also removed proto file changes (will open a separate PR in a moment).

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -934,85 +932,105 @@ impl Drop for LocalShard {
-    cutoff_clocks: Arc<Mutex<ClockMap>>,
-    highest_clocks_path: PathBuf,
-    cutoff_clocks_path: PathBuf,
+    newest_observed_clocks: Arc<Mutex<ClockMap>>,
+    oldest_resolvable_clocks: Arc<Mutex<ClockMap>>,
```
I like short ones better, maybe even: `newest`, `oldest`

We can explain details in comments if we like.

timvisee commented on lib/collection/src/shards/local_shard/shard_ops.rs:
```diff
@@ -247,15 +247,21 @@ impl ShardOperation for LocalShard {
                 sender: callback_sender,
                 wait,
             }));
-            drop(wal_lock);
```
I put it here so we don't make the same mistake again :smile:

ffuugoo replied:
Ok, I kinda get your point now, but, IMO, it's not very effective and this part simply requires an explanation comment (with a lot of exclamation marks in it).

It's still super easy to just look at this code, see the `drop`, think "WTF is this here for?" and just remove the `drop` and refactor the `wal_lock` away same way we did earlier.

If we add the comment, then it will explain that the `drop` is *intentional*, and we *have to* hold the `wal_lock` for the duration of `send`... but at this point there's not much use in doing explicit `drop`, IMO.

timvisee replied:
If I see a `drop` I always think: it must be there for a reason, lets be very careful about it.

Comments don't do the same to me while refactoring.

But anyway, fine with both.

ffuugoo replied:
Hmmm...... I usually do as well, but in this case it's _kinda_ obvious that it's the end of the scope, so I usually assume it's just a leftover from a refactoring or something.

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -934,85 +932,105 @@ impl Drop for LocalShard {
     pub async fn delete_data(shard_path: &Path) -> CollectionResult<()> {
-        let highest_path = Self::highest_clock_map_path(shard_path);
-        let cutoff_path = Self::cutoff_clock_map_path(shard_path);
+        let highest_path = Self::newest_observed_clock_map_path(shard_path);
+        let cutoff_path = Self::oldest_resolvable_clock_map_path(shard_path);
```
Update variable names

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -934,85 +932,105 @@ impl Drop for LocalShard {
     pub async fn move_data(from: &Path, to: &Path) -> CollectionResult<()> {
-        let highest_path_from = Self::highest_clock_map_path(from);
-        let cutoff_path_from = Self::cutoff_clock_map_path(from);
+        let highest_path_from = Self::newest_observed_clock_map_path(from);
+        let cutoff_path_from = Self::oldest_resolvable_clock_map_path(from);
```
Update variable names

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -934,85 +932,105 @@ impl Drop for LocalShard {
     pub async fn copy_data(from: &Path, to: &Path) -> CollectionResult<()> {
-        let highest_path_from = Self::highest_clock_map_path(from);
-        let cutoff_clock_map_path = Self::cutoff_clock_map_path(from);
+        let highest_path_from = Self::newest_observed_clock_map_path(from);
+        let cutoff_clock_map_path = Self::oldest_resolvable_clock_map_path(from);
```
Update variable names

timvisee commented on lib/collection/src/shards/local_shard/mod.rs:
```diff
@@ -934,85 +932,105 @@ impl Drop for LocalShard {
-    pub async fn store(&self) -> CollectionResult<()> {
-        self.cutoff_clocks
+    pub async fn store(&self, shard_path: &Path) -> CollectionResult<()> {
+        // TODO: Is order important here!? If it is, it should be documented with a comment!
```
My reasoning was that the oldest recoverable are more important than the newest seen with regards to consistency.

But I'm not sure if that's actually true.

Before I locked both together but that would be a potential foot gun for a deadlock. I left a comment regarding this, but it's gone for some reason.

timvisee replied:
I think we can remove this TODO.

timvisee commented on lib/collection/src/wal_delta.rs:
```diff
@@ -154,10 +176,10 @@ impl RecoverableWal {
 /// If `None` - the remote WAL is already equal, and we don't have to send any records.
 /// If `Err` - no delta can be resolved.
 fn resolve_wal_delta(
+    operations: impl DoubleEndedIterator<Item = (u64, Option<ClockTag>)>,
```
This one is potentially dangerous, because we say nothing about the order we expect.

timvisee reviewed (commented):
Some comments (and :pick:s) on a quick pass.

ffuugoo commented:
Ok, so the failing CI was my own fault (obviously, lol). It's fixed now. I think we can give it one last pass and merge.

timvisee commented on lib/collection/src/wal.rs:
```diff
@@ -110,8 +110,35 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
+        // concise `from..=self.last_index()`, because if the WAL is empty, `Wal::last_index`
+        // returns `Wal::first_index`, so we end up with `1..=1` instead of an empty range. 😕
+        //
+        // TODO: Refactor `SerdeWal::last_index`/`Wal::last_index` to return `Option<u64>`? 🤔
```
:+1:

ffuugoo replied:
I'd just remove the `TODO` then. Don't feel much like fixing it now, especially, because then `SerdeWall::last_index` and `Wal::last_index` will behave different.

timvisee commented on lib/collection/src/wal.rs:
```diff
@@ -251,6 +240,8 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
     /// If the index is lower than `first_index`, it means we have already acknowledged it but we
     /// are still holding it in a closed segment until it gets truncated.
+    //
+    // TODO: Rename to `first_acknowledged_index`? Or to something else more meaningful than "closed"? 🤔
```
`first_acknowledged_index` fine by me, I don't have a strong opinion on either. I do find both names still a bit confusing though.

I choose `closed` at first because we use the same terminology in other places.

ffuugoo replied:
Decided to ignore this for now, cause I realized what confuses me about the whole `first_index`/`first_closed_index`, is that it _kinda_ makes more sense to give a more specific name to `first_index`, instead of to `first_closed_index`.

Like, `first_closed_index` *is* **the first** index of WAL. And the `first_index` is, like, some semi-arbitrary index somewhere in the WAL. I don't want to dive deeer into it RN, so I leave it as-is. (Just added it to my personal list of Qdrant `TODO`s. :D)

timvisee replied:
Under normal circumstances you'd only use the logical (open) part of the WAL. So in that case, `first_index` probably makes sense. Its only with WAL diff that we want to read over the closed bit too, just because we have that information. Otherwise we'd significantly limit our diffing possibilities.

What about naming it `first_{open,closed}_index()` for now? That may prevent some confusion because you then have to choose a specific method.

ffuugoo commented on lib/collection/src/shards/local_shard/clock_map.rs:
```diff
@@ -204,66 +201,73 @@ impl RecoveryPoint {
-            if clock_tag.clock_tick <= *tick {
+    /// Remove a clock referenced by the clock tag from this recovery point, if the clock is
+    /// *newer or equal* than the tick in the tag.
+    pub fn remove_clock_if_newer_than_or_equal_to_tag(&mut self, tag: ClockTag) {
```
I don't see how it is clear. The argument name does not change the fact, that you can read "if newer or equal" both ways. What are we comparing to what? `clock > tag` or `tag > clock`?

It does make more sense to read "left to right", e.g., `clock > tag`, but, like, do you *absolutely always* do the most logical thing, or are you, like, human? :D

In the end, you *have to* go read the documentation/code to verify that it is indeed `clock > tag`. Using `newer_than_or_equal_to_tag` clarifies `clock > tag`. It is clear to you, because you've been working with it for the last two weeks.

timvisee replied:
`remove_ge_tag`?

The only reasonable assumption I can make about that is: we remove clocks, if greater or equal to the given tag.

timvisee replied:
Or `retain_old_clocks(tag)` with a comment describing the exact condition.

ffuugoo replied:
> `remove_ge_tag`?
> The only reasonable assumption I can make about that is: we remove clocks, if greater or equal to the given tag.

You still can read it both ways. Like, it literally reads as "remove greater or equal tag". Which is a lot like "remove if *tag* is greater of equal". Why are you so against the conjunctions? Just add `remove_ge_to_tag`, and it's clear. Though, I'd still prefer fully verbose one.

> Or `retain_old_clocks(tag)` with a comment describing the exact condition.

That's exactly what I *don't* want. I don't want to *have to* go read the comment. I want it to be obvious when you just read the method name. :/
